### PR TITLE
Share Ext: Added notification support to the share extension

### LINKED ISSUE #8146: Share Extension: Notifications for success & errors
screen shot 2017-11-14 at 8 43 23 amscreen shot 2017-11-14 at 8 43 09 am

* Includes success conditions as well as background errors
* Uses same notification UI as async
* Retry option for errors
* Create/edit post option for success
* Immediate errors should be displayed within app extensions as appropriate (e.g. can't fetch tags)

Depends on [async functionality](https://github.com/wordpress-mobile/WordPress-iOS/projects/10).

/cc @SergioEstevao @frosty

### PR DESCRIPTION
**Pull request by bummytime**
This PR adds notification support to the share extension. Because of the non-deterministic nature of _either_ WPiOS _or_ the extension getting the "did complete" callback, I needed to fire off notifs from 2 different places:
* `ExtensionNotificationManager.swift` (via share extension)
* `ShareExtensionSessionManager.swift` (via WPiOS)

test2
test3

Two things to note here: 
1. The retry action is not implemented in this PR and will be part of the retry work itself.
2. IF the share extension's callback handles the post response...in-app notifs will not work. Perhaps @frosty could offer up some guidance here with his recent work on all this. :-)

Fixes #8146 

**Test 1:**
1. Open the Photos app and share 1 or 2 images.
2. Don't launch WPiOS.
3. You should see a iOS notification after a few seconds.
4. Tap the notification and it should launch WPiOS

**Test 2:**
1. Open the Photos app and share 1 or 2 images.
2. Quickly launch WPiOS after tapping "Publish" on the share extension.
3. You should see an in-app notification after a few seconds.
4. Tap the "Edit Post" action and verify it opens the post in question.

**Test 3:**
1. Open Safari and share a web page.
2. Don't launch WPiOS.
3. You should see a iOS notification after a few seconds.
4. Tap the notification and it should launch WPiOS.

@frosty Would you mind taking a peek at this? Thank you!!!!

### CODE REVIEW

File: WordPress/WordPressShareExtension/ExtensionNotificationManager.swift [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        let notificationTrigger = UNTimeIntervalNotificationTrigger(timeInterval: delay, repeats: false)
+        let notificationRequest = UNNotificationRequest(identifier: UUID().uuidString, content: notificationContent, trigger: notificationTrigger)
+        UNUserNotificationCenter.current().add(notificationRequest, withCompletionHandler: { error in
+            if let error = error {
+                DDLogError("Unable to add notification request (\(error), \(error.localizedDescription))")
+            }
+        })
+    }
+
+    private static func createUserInfoDict(postID: String, blogID: String) -> [String: Any] {
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ExtensionNotificationManager.swift]: Total nitpick, but the [Swift API guidelines](https://swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage) prefer `make` over `create`

Comment (bummytime) [reply WordPress/WordPressShareExtension/ExtensionNotificationManager.swift]: Done!

**Comment by bummytime** [comment]
@frosty thanks for the review!

> In the case of Photos, I couldn't seem to get the notifications to appear if WPiOS wasn't running at all. If I launched WPiOS, went back to Photos and then shared, the notifications would come in. But not if I quit WPiOS first. Have you seen this before?

I have not seen this but I will look into it tonight!

**Comment by bummytime** [comment]
Also,

> I think personally I'd prefer no periods.

Ditto. 😃

**Comment by frosty** [comment]
@bummytime So am I correct in thinking the issue is that you can't fire an in-app notification (notice) from the share extension – it won't appear in WPiOS? I guess that's because they're in separate processes... Here's one suggestion:

* Instead of trying to fire notices from the share extension, always fire local notifications.
* In WPiOS, in `InteractiveNotificationsManager`, we can implement `userNotificationCenter(_:willPresent:withCompletionHandler:)`. 
* This gets called whenever local notifications are received by the app in the foreground. Normally, you can tell the system how to handle these notifications – for example, you can call the completion handler with a value of `.alert`, and the system will display a normal notification alert within the app.
* In this method, we can detect if it's a local notification we're interested in (which it should be), and immediately dispatch a Notice using the notification's content so it shows up as a notice.
* We'd then just call the completion handler with an empty return value `[]` so the system ignores it.

Does that make sense?

**Comment by bummytime** [comment]
Thanks @frosty! That sounds like a solid plan. Let me imp that this morning and see if it works.

**Comment by bummytime** [comment]
@frosty, see if this latest round of changes works for you. In my testing it seems to have helped.

File: WordPress/WordPressShareExtension/ExtensionNotificationManager.swift [hunk 2]
```diff
@@ -59,10 +61,12 @@ class ExtensionNotificationManager {
         })
     }
 
-    private static func makeUserInfoDict(postID: String, blogID: String) -> [String: Any] {
+    private static func makeUserInfoDict(postUploadOpID: String, postID: String, blogID: String) -> [String: Any] {
         var userInfo = [String: Any]()
+        userInfo[ShareNoticeUserInfoKey.postUploadOpID] = postUploadOpID
         userInfo[ShareNoticeUserInfoKey.postID] = postID
         userInfo[ShareNoticeUserInfoKey.blogID] = blogID
+        userInfo[ShareNoticeUserInfoKey.originatedFromAppExtension] = true
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ExtensionNotificationManager.swift]: Nice idea using this key!

**Comment by frosty** [comment]
So I've taken this for a spin again, and the following situations work really well and are really reliable:

* App is open in the background, and photos or web links are shared. Local notification is presented on successful upload.
* App is not open in the background and web links are shared. Local notification is presented on successful upload.
* Photos or web links are shared and the app is then opened in the foreground before uploads complete. In-app notice is displayed.

I'm still seeing the issue I had before where if I quit the app from the multitasker and share photos, I don't get a notification. Unless perhaps I need to leave it longer. One time when testing I left it for about 2 minutes, and then re-opened the app – at which point I was shown a notice:

img_f7cb07fb68d3-1

It may be that this is just going to be a limit of share extensions and there's not much else we can do.

Also, failure notices aren't being displayed because their category hasn't been added to `InteractiveNotificationsManager` – wasn't sure if this was intentional or not?

**Comment by bummytime** [comment]
> It may be that this is just going to be a limit of share extensions and there's not much else we can do.

I am wondering the same thing at this point. Perhaps I could just make another issue for us to track it?

> Also, failure notices aren't being displayed because their category hasn't been added to InteractiveNotificationsManager – wasn't sure if this was intentional or not?

Not intentional. I will add the category!

> uses full punctuation (the title / messages end with periods), whereas the async notifications don't. Should we standardise here? I think personally I'd prefer no periods.

I will also fix this. :-)

**Comment by bummytime** [comment]
Ready for (hopefully) a final look @frosty! Thank you!!

**Comment by frosty** [comment]
So I'm unable to actually get an upload to fail, but I think this is a quirk of the share extension system... I switched into airplane mode before posting, left it for a couple of minutes and still no failure notice. As soon as I turned airplane mode back off, the post uploaded successfully and I saw a success notification. 

The code looks good, in the event of a failure, so I suggest we should just go ahead and :shipit:

**Comment by bummytime** [comment]
Thank you @frosty !!

**Comment by folletto** [comment]
>  I noticed that the text for these notifications uses full punctuation (the title / messages end with periods), whereas the async notifications don't. Should we standardise here? I think personally I'd prefer no periods.

My choice is for standardization - no preference for which.
It seems to me on iOS notifications go without, but... I'm ok with your choice. :)

_(Sorry for the late reply)_

**Pull request merged**